### PR TITLE
fixing z plane duplicate removal

### PR DESCRIPTION
**Pull request by seichhorn**
In using this I encountered a couple edge cases relating to having none of a particular barcode. These have been handled now. I also was referencing z positions in the barcodedb wrong, I forgot we use z index not z position, and fixed this now.

### CODE REVIEW

File: merlin/analysis/filterbarcodes.py [hunk 1]
```diff
@@ -395,7 +395,9 @@ def _run_analysis(self, fragmentIndex):
         bcDatabase = self.get_barcode_database()
         currentBarcodes = decodeTask.get_barcode_database()\
             .get_barcodes(fragmentIndex)
-        currentBarcodes = self.remove_z_duplicate_barcodes(currentBarcodes)
+
```
Comment (emanuega) [inline merlin/analysis/filterbarcodes.py:398]: This seems like it could change the "misidentification rate." Would it be possible to incorporate this into the calculation of the adaptive filter? I don't know the cleanest way to do this. Maybe the GenerateAdaptiveFilter task needs to know about the AdaptiveFilterBarcodes task so that it can read the value of remove_z_duplicate_barcodes. I don't want them to both have this parameter since it doesn't make sense to set this parameter indepedently.

File: merlin/util/barcodefilters.py [hunk 2]
```diff
@@ -72,21 +75,26 @@ def remove_zplane_duplicates_single_barcodeid(barcodes: pd.DataFrame,
...
-        for compZ in zToCompare:
-            queryBC = barcodes[barcodes['z'] == compZ]
-            dist, idx = tree.query(queryBC.loc[:, ['x', 'y']].values, k=1,
-                                   distance_upper_bound=maxDist)
-            currentHits = treeBC.index.values[idx[np.isfinite(dist)]]
-            comparisonHits = queryBC.index.values[np.isfinite(dist)]
-            graph.add_edges_from(list(zip(currentHits, comparisonHits)))
-    connectedComponents = [list(x) for x in
-                           list(nx.connected_components(graph))]
+        if len(treeBC) == 0:
```
Comment (emanuega) [inline merlin/util/barcodefilters.py:82]: Can you add a new test for the edge cases you are fixing here?

File: merlin/analysis/filterbarcodes.py (hunk 1 above)
Comment (seichhorn) [reply merlin/analysis/filterbarcodes.py:398]: Yeah, I've thought about this. It does change the apparent misidentification rate for a given experiment, in my test cases it tended to increase the misidentification rate approximately proportional to the fraction removed as duplicates (i.e. if you removed 20% of your barcodes as apparent duplicates, the misidentification rate tended to go up about 20%), since it seems like blanks are depleted among the zplane duplicates. 

It would be most straightforward to address the duplicated barcodes at the decoding step, not the filtering step, and generate the adaptive threshold on the cleaned set of barcodes. It would add about a minute to the analysis of each z plane this way, which isn't bad. Would you prefer I implement it at decoding?

Comment (emanuega) [reply merlin/analysis/filterbarcodes.py:398]: That could work too. At first it seemed challenging since we decode and write one z plane at a time and wouldn't have enough information to remove duplicates, but I imagine we could read all the barcodes back in at the end and rewrite the barcodes with duplicates removed. Is this what only takes about a minute? If so this could be a good solution.

Comment (seichhorn) [reply merlin/analysis/filterbarcodes.py:398]: Yeah, I figured I would just read in all the barcodes at the end and overwrite the original result. With a dataset that has 285 barcodes and 6 z planes, it takes about 3 minutes to remove duplicates when run on the full decoding result of a FOV. I think it should scale pretty linearly with number of decoded barcodes and the number of z planes you ask it to test against.

**Comment by seichhorn** [comment]
@emanuega I think this is ready now

File: merlin/analysis/decode.py [hunk 3]
```diff
@@ -20,6 +21,14 @@ def __init__(self, dataSet: dataset.DataSet, parameters=None,
                  analysisName=None):
         super().__init__(dataSet, parameters, analysisName)
 
+        if 'remove_z_duplicated_barcodes' not in self.parameters:
```
Comment (emanuega) [inline merlin/analysis/decode.py:24]: Can this move to the Decode task? Many other tasks, such as the filter tasks are BarcodeSavingParallelAnalysisTasks but don't need these parameters.
